### PR TITLE
store: reuse chunks map

### PR DESCRIPTION
**Pull request by GiedriusS**
Reuse chunks map instead of creating a new one each time. This is a hot path and shows up in profiles.

### CODE REVIEW

File: pkg/store/proxy_merge.go [hunk 1]
```diff
@@ -101,15 +104,15 @@ func (d *responseDeduplicator) Next() bool {
...
+		d.bufferedResp = append(d.bufferedResp, d.chainSeriesAndRemIdenticalChunks(d.bufferedSameSeries))
 		d.prev = s
 
 		return true
 	}
 }
 
-func chainSeriesAndRemIdenticalChunks(series []*storepb.SeriesResponse) *storepb.SeriesResponse {
-	chunkDedupMap := map[uint64]*storepb.AggrChunk{}
+func (d *responseDeduplicator) chainSeriesAndRemIdenticalChunks(series []*storepb.SeriesResponse) *storepb.SeriesResponse {
```
Comment (yeya24) [inline pkg/store/proxy_merge.go:114]: One question unrelated to this PR. Is the main usecase of `chainSeriesAndRemIdenticalChunks` to dedup chunks from Receiver? How likely those receiver replicas have exact the same chunk

Comment (GiedriusS) [reply pkg/store/proxy_merge.go:114]: Yeah, it depends on the replication factor. We have RF=2 so there's always at least one extra copy.

**Review by niaurys — APPROVED** [review]

**Review by yeya24 — APPROVED** [review]
Thanks.

Just curious, how much does this codepath show up in your profile? For us, it never show up.

In the future, I'd love to see a feature flag to just disable this chunks deduplication functionality since our usecase never needs it. We only query blocks that already deduplicated from store gateway.

**Comment by GiedriusS** [comment]
> Thanks.
> 
> Just curious, how much does this codepath show up in your profile? For us, it never show up.
> 
> In the future, I'd love to see a feature flag to just disable this chunks deduplication functionality since our usecase never needs it. We only query blocks that already deduplicated from store gateway.

In our profiles this function itself is responsible for ~5% of allocs

**Pull request merged**